### PR TITLE
Fix module reloading

### LINKED ISSUE #21: module is not reloadable
I use master

```
tarantool> _ = require('tuple.merger')
---
...

tarantool> package.loaded['tuple.merger'] = nil
---
...

tarantool> _ = require('tuple.merger')
---
- error: 'tuple.merger/postload.lua:34: cannot change a protected metatable'
...
```

Seems it's the same as in https://github.com/tarantool/tuple-keydef/issues/9

### PR DESCRIPTION
**Pull request by olegrok**
This patchset fixes module hotreload (in the same way as in https://github.com/tarantool/tuple-keydef/pull/10). See commit messages for details.

Closes #21

### CODE REVIEW

**Comment by Totktonada** [comment]
How about the alternative way (https://github.com/tarantool/tuple-keydef/pull/15)?

**Comment by olegrok** [comment]
> How about the alternative way ([tarantool/tuple-keydef#15](https://github.com/tarantool/tuple-keydef/pull/15))?

Yes, of course. New version is pushed.

**Comment by Totktonada** [comment]
There were minor updates in https://github.com/tarantool/tuple-keydef/pull/16 (using of `lengthof()` + a bit more comments), if you want to keep the code more or less same. Anyway, code updates looks okay for me. I'll test it and, if everything is okay, will approve.

**Comment by olegrok** [comment]
> There were minor updates in [tarantool/tuple-keydef#16](https://github.com/tarantool/tuple-keydef/pull/16) (using of `lengthof()` + a bit more comments), if you want to keep the code more or less same. Anyway, code updates looks okay for me. I'll test it and, if everything is okay, will approve.

There is no "lengthof" declaration in this module. If you don't mind I add just `size_t len = sizeof(meta) / sizeof((meta)[0]);`. I'm not sure that we should introduce new file with dependency if some part if used only once.

The rest of comments I moved to this patch.

File: src/merger/merger.c [hunk 1]
```diff
@@ -1256,15 +1267,17 @@ luaopen_tuple_merger(struct lua_State *L)
...
 		{"new_buffer_source", lbox_merger_new_buffer_source},
 		{"new_table_source", lbox_merger_new_table_source},
 		{"new_tuple_source", lbox_merger_new_tuple_source},
 		{"new", lbox_merger_new},
 		{NULL, NULL}
 	};
-	luaL_register(L, "tuple.merger", meta);
+	size_t len = sizeof(meta) / sizeof(meta[0]);
+	lua_createtable(L, 0, len);
+	luaL_register(L, NULL, meta);
```
Comment (Totktonada) [inline src/merger/merger.c:1280]: It'll count `{NULL, NULL}`, so we should substract 1.

File: test/merger-test.lua [hunk 2]
```diff
@@ -765,4 +765,28 @@ for _, input_type in ipairs({'buffer', 'table', 'tuple'}) do
     end
 end
 
+test:test('hotreload', function(test)
```
Comment (Totktonada) [inline test/merger-test.lua:768]: I would catch the 'declared_ctype' [test](https://github.com/tarantool/tuple-keydef/blob/109c91361b9d84a70e6dafd79afadd60eca136ed/test/hot_reload.test.lua#L34-L59) as well. However I can do it as follow up to don't tire you.

**Comment by Totktonada** [comment]
`require('tuple.merger').new_source_frombuffer` and `require('tuple.merger').new_source_frombuffer` are gone after reloading.

File: src/merger/merger.c (hunk 1 above)
Comment (olegrok) [reply src/merger/merger.c]: Yes, thanks. Fixed.

File: test/merger-test.lua (hunk 2 above)
Comment (olegrok) [reply test/merger-test.lua]: Yes, moved test hotreload suite from tuple-keydef module into separate file.

**Comment by olegrok** [comment]
> `require('tuple.merger').new_source_frombuffer` and `require('tuple.merger').new_source_frombuffer` are gone after reloading.

Good catch. Fixed via passing "first_load" flag into postload.lua.

File: src/postload.lua [hunk 3]
```diff
@@ -1,3 +1,10 @@
+-- This script is executed only on the first load of the module.
+--
+-- It is NOT executed after hot reload (when the package.loaded
+-- entry is removed and require is called once again).
```
Comment (Totktonada) [inline src/postload.lua]: It seems, you forgot to adjust this comment.

File: test/hotreload-test.lua [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        'new_source_fromtable',
+        'new_source_frombuffer',
+        'new_tuple_source',
+    }
+    test:plan(#methods)
+
+    for _, m in ipairs(methods) do
+        test:ok(tuple_merger[m] ~= nil, ('%s is present'):format(m))
+    end
+end
```
Comment (Totktonada) [inline test/hotreload-test.lua:24]: The comment is about instance methods, but the test verifies module functions. It seems, we need two separate checks for them.

**Comment by Totktonada** [comment]
Please, look a few minor comments above. Everything else looks good for me.

File: src/postload.lua (hunk 3 above)
Comment (olegrok) [reply src/postload.lua]: Yes, thanks. Fixed.

File: test/hotreload-test.lua (hunk 4 above)
Comment (olegrok) [reply test/hotreload-test.lua]: Agree. Added.

File: test/hotreload-test.lua [hunk 5]
```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env tarantool
+
+local tap = require('tap')
+local ffi = require('ffi')
+local key_def_lib = require('tuple.keydef')
+
+-- Presence of the methods confirms that correct metatype is set
+-- for the <struct tuple_merge_source> ctype.
```
Comment (Totktonada) [inline test/hotreload-test.lua:8]: Sorry for nitpicking, but presence of module functions (BTW, it is not quite correct to call them methods) just verifies that they're here. While presence of instance methods proves that the metatype is correct.

Comment (olegrok) [reply test/hotreload-test.lua]: Sorry for delay. Yes, thanks for notice. This comment is related to function below. I moved to correct place and added appropriate comment for this function.

**Review by Totktonada — APPROVED** [review]
LGTM.

**Pull request merged**